libkmod/libkmod-builtin.c

```diff
@@ -1,333 +1,193 @@
+	ssize_t count, n;
 
-	iter->nstrings = 0;
-	offset = pos = iter->next;
+	for (count = 0; count < INTPTR_MAX;) {
+		char *dot, *line;
 
-	while (offset < iter->size) {
-		char *dot;
-		off_t len;
-
-		offset = get_string(iter, pos, &line, &linesz);
-		if (offset <= 0) {
-			if (offset)
-				ERR(iter->ctx, "get_string: %s\n", strerror(errno));
-			pos = iter->size;
+		n = get_string(info);
+		if (n == -1) {
+			if (!feof(info->fp)) {
+				count = -errno;
+				ERR(info->ctx, "get_string: %s\n", strerror(errno));
+			}
 			break;
 		}
 
+		line = info->buf;
 		dot = strchr(line, '.');
-		if (!dot) {
-			ERR(iter->ctx, "kmod_builtin_iter_next: unexpected string without modname prefix\n");
-			pos = iter->size;
-			break;
+		if (dot == NULL) {
+			count = -EINVAL;
+			ERR(info->ctx, "get_strings: "
+			    "unexpected string without modname prefix\n");
+			return count;
 		}
-
-		len = dot - line;
-
-		if (!modname) {
-			modname = strdup(line);
-			modlen = len;
-		} else if (modlen != len || strncmp(modname, line, len)) {
+		if (strncmp(line, modname, modlen) || line[modlen] != '.') {
+			/*
+			 * If no string matched modname yet, keep searching.
+			 * Otherwise this indicates that no further string will
+			 * match again.
+			 */
+			if (count == 0)
+				continue;
 			break;
 		}
-
-		iter->nstrings++;
-		pos = offset;
+		if (!strbuf_pushchars(buf, dot + 1) ||
+		    !strbuf_pushchar(buf, '\0')) {
+			count = -errno;
+			ERR(info->ctx, "get_strings: "
+			    "failed to append modinfo string\n");
+			return count;
+		}
+		count++;
 	}
 
-	iter->pos = iter->next;
-	iter->next = pos;
-
-	free(modname);
+	if (count == INTPTR_MAX) {
+		count = -ENOMEM;
+		ERR(info->ctx, "get_strings: "
+		    "too many modinfo strings encountered\n");
+		return count;
+	}
 
-	return (iter->pos < iter->size);
+	return count;
 }
 
-static bool kmod_builtin_iter_get_modname(struct kmod_builtin_iter *iter,
-				char modname[static PATH_MAX])
+static char **strbuf_to_vector(struct strbuf *buf, size_t count)
 {
-	int sv_errno;
-	char *line, *dot;
-	size_t linesz, len;
-	off_t offset;
-
-	if (iter->pos == iter->size) {
-		sv_errno = EINVAL;
-		goto fail;
-	}
-
-	line = NULL;
+	/* size required for string vector + terminating NULL */
+	const size_t vecsz = sizeof(char *) * (count + 1);
+	char **vector;
+	char *s;
+	size_t n;
 
-	offset = get_string(iter, iter->pos, &line, &linesz);
-	if (offset <= 0) {
-		sv_errno = errno;
-		if (offset)
-			ERR(iter->ctx, "get_string: %s\n", strerror(errno));
-		goto fail;
+	/* make sure that vector and strings fit into memory constraints */
+	if (SIZE_MAX / sizeof(char *) - 1 < count ||
+	    SIZE_MAX - buf->used < vecsz) {
+		errno = ENOMEM;
+		return NULL;
 	}
 
-	dot = strchr(line, '.');
-	if (!dot) {
-		sv_errno = EINVAL;
-		ERR(iter->ctx, "kmod_builtin_iter_get_modname: unexpected string without modname prefix\n");
-		goto fail;
-	}
-
-	len = dot - line;
+	vector = realloc(buf->bytes, vecsz + buf->used);
```

Comment:
instead of forcing a realloc, should we rather grow strbuf? It may be that we still have room and don't actually need the extra alloc.

Reply:
I've used the same `realloc` which would occur in `strbuf_steal`, i.e. even if we have enough memory, don't waste additional unneeded one.